feat: add documentation for navigation capabilities in the UI

matthew-chirgwin opened the pull request:
- Adds documentation describing how introspection, navigation and routing will work in the Strimzi-ui

Closes: #12

matthew-chirgwin committed:
feat: Add Navigation, routing and introspection documentation

- adds documentation describing how routing, navigation, and
backend introspection will operate

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on client/Contexts/Introspect/README.md:
```diff
@@ -0,0 +1,79 @@
+    Topic: {
+        type: 'Topic', // the GraphQL type name
+        fields: {
+            name: GQL_STRING, // the expected GraphQL type - could be string or object - which is checked if is available and thus query-able
```
If we have an "unreliable" schema (hence the introspection) - would it be simpler to have a translation layer?

As in - components call `buildQuery` with the "complete" query, and fields are filtered back out? And the same way with a `parseResults`, it results in an object with `null`s for bits that are missing?

My concern is that the components now have to build composed queries based off fields/capabilities.

What would be great is if introspection told the component what "capabilities" can be used (and those control features) - but the data in/out was just filtered out so that you just handled null if something you needed was missing?

matthew-chirgwin replied:
> What would be great is if introspection told the component what "capabilities" can be used (and those control features)

That is the intent here - this model descirbes what is expected, which can then be checked with flags being enabled /disabled for example. The only usage so far is to show/hide whole pages, but this could be used at a feature flag level as well.

I must admit, I have not considered a partial schema case (and returning nulls/empty values) for a query. Let me look into this and see what would need to change to support it.

nictownsend commented on client/Bootstrap/Navigation/README.md:
```diff
@@ -0,0 +1,271 @@
+
+## Route configuration processing
+
+As mentioned, this component takes in [configuration](../../Pages/README.md) which maps [`Panels`](../../Panels/README.md) to routes, along with other supporting metadata, such as a translatable page names, what backend services are required to show the page, and what authorization(s) a user may need to see it. This configuration is processed via the Navigation's model, and will result in an obje ...
```
`home/topic/1/partitions` - what if that was to render the topic page, but with a slide out panel enabled?

Might be me thinking too hard about this - but it looks like this is tying us such that a path maps to a single component.

And is there a way to add dynamic properties (i.e pass through something like a query param or a route param) - so that the component is unaware of the history or route objects?

matthew-chirgwin replied:
For elements like that, I would consider a query parameter perhaps. The job of the Navigation is to render routes based on configuration, and pass contextual information (path parameters, query parameters) to the configured component. What the component does with those pieces of information are not in my opinion the responsibility of the navigation component.

To your second point, I have prototyped and validated the ability to pass things like path paramters, query parameters via a property to the component on render. I did leave it as a todo for implementation, but I will update the PR with more details on this now.

nictownsend commented on client/Pages/README.md:
````diff
@@ -0,0 +1,103 @@
+
+The schema used to define pages in the Strimzi-ui is designed around capability, and the scenarios those capabilities are presented in. The exported object should be named to reflect the capability it provides, and the metadata contained to specify the `Panel` component used to provide it (`contentComponent`), and where (and how) it should be presented (`contexts`). The below example (pseudo) sha ...
+
+```
````
Is this a duplication of the content above?

matthew-chirgwin replied:
By above, do you mean in the Pages README? If so, I see the text above this section as more of a summary, and this as more detail about the structure itself.

nictownsend commented on client/Bootstrap/Navigation/README.md:
```diff
@@ -0,0 +1,271 @@
+                },
+                requiresMinimum: {
+                    backendSupportFor: {
+                        Group: { READ }
```
Where are these constants defined? And what about subscription support?

matthew-chirgwin replied:
Did you see https://github.com/matthew-chirgwin/strimzi-ui/blob/12-routing-nav/docs/Architecture.md#entity-model ? I do not mention where the common constants come from, so will add that, but the idea is that the 'abilities' on the entity could be anything, and that they are contextual, known and used/referenced by the code that needs them.

matthew-chirgwin committed:
refactor: address review comments

- clarify navigation state passed to components on render
- remove id from pages metadata
- add reference to entity model constants file

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented:
@dlabaj / @nictownsend / @scholzj thank you all for the reviews/comments. I have just pushed an update to address them. Please let me know if you have any further thoughts.

@nictownsend re 'filling in blanks' around queries - I think the introspection result could be used to help guide a query helper (or similar) to fill in blanks if required. I do err however that it may then cause other issues (if users of the query result do not handle nulls etc well). Would like to discuss with you further on it though.

nictownsend reviewed (changes requested):
LGTM except for one small point re `backendSupportFor` - how can you articulate the need for a subscription vs query?

matthew-chirgwin commented:
> LGTM except for one small point re `backendSupportFor` - how can you articulate the need for a subscription vs query?

Where a subscription is required, this would be an additional `backendSupportFor` key to check for it. At which point the callback for that key would be proivided the schema. There is an example of this here: https://github.com/matthew-chirgwin/strimzi-ui/blob/12-routing-nav/docs/Architecture.md#entity-model (`liveUpdates`)

nictownsend reviewed (approved):
Thanks, LGTM for laying groundwork
